### PR TITLE
Templating: global/system variables should be properly replaced in templated values.

### LINKED ISSUE #27335: PanelHeader: $__dashboard.name does not work in panel title
**What happened**:
I'm using the $__dashboard in a panel, but instead of the dashboard name I see undefined
image

**What you expected to happen**:
That the dashboard name will appear.

**How to reproduce it (as minimally and precisely as possible)**:
Create a panel and add $__dashboard to the title.

**Anything else we need to know?**:

**Environment**:
- Grafana version: 7.1.5 (but I don't think it's a regression)
- Data source type & version: N/A
- OS Grafana is installed on: docker
- User OS & Browser: fedora/chrome
- Grafana plugins: N/A
- Others:

### PR DESCRIPTION
**Pull request by mckn**
**What this PR does / why we need it**:
Looks like we have lost support of using system/global template variables when replacing labels. In the issue referenced in this PR the templating of panel title stopped working when entering `${__dashboard.name}` as the panel title.

**Which issue(s) this PR fixes**:
Fixes #27335

**Special notes for your reviewer**:
The code is kind of weird. In this example we want to use a system/global template variable as the interpolation value but as far as I kan see we never use the `templateSrv.setGlobalVariable` function (it is deprecated) to set the `templateSrv.grafanaVariables`. In the `templateSrv.replace*` function we are still trying to use the `templateSrv.grafanaVariables` list as the source.

I would love to have a look at this together with someone else that knows more about the different scenarios.

### CODE REVIEW

**Review by hugohaggmark — CHANGES_REQUESTED** [review]
I must have missed this function when I moved system variables to Redux state. I don't think we need this function so I think we should deprecate it and have it call the replace function internally or remove it and just use templateSrv.replace instead? That said, I don't know all the special cases here either :)

**Comment by torkelo** [comment]
Tricky, this is slightly different from the normal replace as this replaces with the text instead if value property (So show say All if all are selected , and will show text representation if multiple selected, or text representation if text and value are different)

**Comment by hugohaggmark** [comment]
Thank you for clarifying @torkelo.

**Review by hugohaggmark — COMMENTED** [review]
Don't think using `this.grafanaVariables` is a way forward, but maybe that would become a bigger PR. Would be happy to figure something out together with you @mckn

**Comment by mckn** [comment]
@torkelo me and @hugohaggmark had a quick sync regarding this and came up with a possible way forward.

1. Extend the `replace` with a new format (`:text`) that we register in the format registry.
2. Deprecate the old `replaceText` functionality by adding a deprecation warning. Change so `replaceText` calls `replace` with the `:text` format internally.
3. Implement a `TextFormatter` and add some more tests for the `replace` functionality.

Would this work or do you see any issues with this approach?

**Comment by torkelo** [comment]
I was thinking  the same, that maybe a format could be used so we can share the same replace function. 

TextFormatter could ignore value and return variable.current.text

File: public/app/features/templating/formatRegistry.ts [hunk 1]
```diff
@@ -2,8 +2,14 @@ import kbn from 'app/core/utils/kbn';
 import { Registry, RegistryItem, VariableModel, textUtil, dateTime } from '@grafana/data';
 import { map, isArray, replace } from 'lodash';
 
+export interface FormatOptions {
+  value: any;
+  args: string[];
+  text: string;
+}
+
```
Comment (hugohaggmark) [inline public/app/features/templating/formatRegistry.ts:10]: Nice ⭐

File: public/app/features/templating/formatRegistry.ts [hunk 2]
```diff
@@ -184,13 +190,30 @@ export const formatRegistry = new Registry<FormatRegistryItem>(() => {
...
+        }
+
+        const current = (variable as any)?.current;
+
+        if (typeof current?.value === 'string') {
+          return current.value;
+        }
+        return current.text;
+      },
+    },
```
Comment (hugohaggmark) [inline public/app/features/templating/formatRegistry.ts:216]: nit: is this clearer?
```suggestion
       if(!current) {
         return '';
       }
       
       const {value, text} = current;

        if (typeof value === 'string') {
          return value;
        }
        return text;
      },
    },
```

File: public/app/features/templating/template_srv.ts [hunk 3]
```diff
@@ -275,30 +301,31 @@ export class TemplateSrv implements BaseTemplateSrv {
...
+      if (format) {
+        return match.replace(`${variable}:${format}`, `${variable}:text`);
+      }
 
-      return typeof value === 'string' ? value : variable.current.text;
+      if (match.indexOf(`$${variable}`) > -1) {
+        return match.replace(`$${variable}`, `\${${variable}:text}`);
+      }
+      return match.replace(`${variable}`, `${variable}:text`);
     });
```
Comment (hugohaggmark) [inline public/app/features/templating/template_srv.ts:326]: Something about this feels hacky

Comment (mckn) [reply public/app/features/templating/template_srv.ts:326]: Yeah, it kind of is :)

**Review by hugohaggmark — CHANGES_REQUESTED** [review]
Skilled work and I think we're almost there. When using multi value variable in the title the output should be exactly as the Picker, i.e. `B + C` and not `B, C`.

Current version:
image

Grafana 6.7.4:
image

In an follow-up PR we need to:

- [x] Add new format to docs
- [x] Update the gdev dashboard devenv/dev-dashboards/feature-templating/global-variables-and-interpolation.json
- [x] Update the e2e test that runs on that test dashboard
- [x] Picker and formatter should use same function

**Comment by mckn** [comment]
> Skilled work and I think we're almost there. When using multi value variable in the title the output should be exactly as the Picker, i.e. `B + C` and not `B, C`.
> 
> Current version:
> image
> 
> Grafana 6.7.4:
> image
> 
> In an follow-up PR we need to:
> 
> * [ ]  Add new format to docs
> * [ ]  Update the gdev dashboard devenv/dev-dashboards/feature-templating/global-variables-and-interpolation.json
> * [ ]  Update the e2e test that runs on that test dashboard
> * [ ]  Picker and formatter should use same function

If you do a pull and try again it should be fixed. But we need to do all the things that you added on the list above.

File: public/app/features/templating/template_srv.test.ts [hunk 4]
```diff
@@ -95,6 +95,20 @@ describe('templateSrv', () => {
...
+      });
+      expect(target).toBe('this.mupp.filters');
+    });
+
+    it('should replace ${test.name} with variable text', () => {
+      const target = _templateSrv.replaceWithText('this.${test.name}.filters', {
+        test: { value: { name: 1 }, text: 'asd' },
+      });
+      expect(target).toBe('this.asd.filters');
+    });
```
Comment (torkelo) [inline public/app/features/templating/template_srv.test.ts:110]: Not sure if this use case exists

File: public/app/features/templating/formatRegistry.ts [hunk 5]
```diff
@@ -184,13 +190,41 @@ export const formatRegistry = new Registry<FormatRegistryItem>(() => {
...
+      description: 'Format variables in their text representation. Example in multi variable scenario A + B + C.',
+      formatter: (options, variable) => {
+        if (typeof options.text === 'string') {
+          return options.text;
+        }
+
+        const current = (variable as any)?.current;
+
+        if (!current) {
+          return '';
```
Comment (torkelo) [inline public/app/features/templating/formatRegistry.ts:212]: can we return value here?

Comment (mckn) [reply public/app/features/templating/formatRegistry.ts:212]: yeah - why not :)

File: public/app/features/templating/formatRegistry.ts [hunk 6]
```diff
@@ -184,13 +190,41 @@ export const formatRegistry = new Registry<FormatRegistryItem>(() => {
...
+        }
+
+        const { value, text } = current;
+
+        if (Array.isArray(value)) {
+          return value.join(' + ');
```
Comment (torkelo) [inline public/app/features/templating/formatRegistry.ts:218]: Should this not just return text? and if text is array do this

Comment (torkelo) [reply public/app/features/templating/formatRegistry.ts:218]: but use text

File: public/app/features/templating/template_srv.test.ts (hunk 4 above)
Comment (torkelo) [reply public/app/features/templating/template_srv.test.ts:110]: this looks wrong to me, value.name only has one representation here (name), the text is on a different level

**Review by hugohaggmark — APPROVED** [review]
That was super fast and works as expected 🌟  and I agree that we should probably fix the list of things I added.

File: public/app/features/templating/template_srv.test.ts [hunk 7]
```diff
@@ -595,6 +609,45 @@ describe('templateSrv', () => {
     });
   });
 
+  describe('replaceWithText can pass all / multi value', () => {
+    beforeEach(() => {
+      initTemplateSrv([
+        {
+          type: 'query',
+          name: 'server',
+          current: { value: ['server1', 'server2'] },
```
Comment (torkelo) [inline public/app/features/templating/template_srv.test.ts:618]: this is missing text representation

File: public/app/features/templating/template_srv.ts [hunk 8]
```diff
@@ -233,8 +255,10 @@ export class TemplateSrv implements BaseTemplateSrv {
 
       if (scopedVars) {
         const value = this.getVariableValue(variableName, fieldPath, scopedVars);
+        const text = this.getVariableText(variableName, fieldPath, scopedVars);
```
Comment (torkelo) [inline public/app/features/templating/template_srv.ts:258]: this is a bit expensive (if there is a field path), would be nice to only have to do this if format is text. 

Maybe we can have a special logic here for format == text, and then pass that as value to this.formatValue

File: public/app/features/templating/template_srv.ts [hunk 9]
```diff
@@ -219,6 +220,27 @@ export class TemplateSrv implements BaseTemplateSrv {
...
+    if (!scopedVar) {
+      return null;
+    }
+
+    if (fieldPath) {
+      const value = this.getFieldAccessor(fieldPath)(scopedVar.value);
+
+      if (typeof value?.text === 'string') {
+        return value.text;
+      }
```
Comment (torkelo) [inline public/app/features/templating/template_srv.ts:234]: I do not think we should support this (we do not currently), and not sure we have a use case for this

File: public/app/features/templating/template_srv.ts (hunk 3 above)
Comment (torkelo) [reply public/app/features/templating/template_srv.ts:327]: not sure I understand this, I though this function was just to call this.replace(format, scopedVars, 'text');

**Review by torkelo — CHANGES_REQUESTED** [review]
Added some comments, can jump on a call later to discuss it but feels like we cannot merge in for beta1

File: public/app/features/templating/template_srv.ts (hunk 3 above)
Comment (hugohaggmark) [reply public/app/features/templating/template_srv.ts:327]: I agree with @torkelo, we should just call this.replace(format, scopedVars, 'text') here instead?

Comment (mckn) [reply public/app/features/templating/template_srv.ts:327]: ahh, I didn't see the forest for all the trees. Thanks 🙏

File: public/app/features/templating/template_srv.ts (hunk 8 above)
Comment (mckn) [reply public/app/features/templating/template_srv.ts:258]: hmm, I can see that we have a cache. Is it expensive to do regardless of it being cached?

File: public/app/features/templating/template_srv.test.ts [hunk 10]
```diff
@@ -95,6 +95,13 @@ describe('templateSrv', () => {
       expect(target).toBe('this.asd.filters');
     });
 
+    it('should replace ${test.name} with scoped text', () => {
```
Comment (torkelo) [inline public/app/features/templating/template_srv.test.ts:98]: this description is no longer valid (it replaces with value)

File: public/app/features/templating/template_srv.ts [hunk 11]
```diff
@@ -275,30 +295,23 @@ export class TemplateSrv implements BaseTemplateSrv {
...
+      const format = fmt2 ?? fmt3;
+      if (!format) {
         return match;
       }
-
-      const value = this.grafanaVariables[variable.current.value];
-
-      return typeof value === 'string' ? value : variable.current.text;
+      return match.replace(`:${format}`, '');
     });
```
Comment (torkelo) [inline public/app/features/templating/template_srv.ts:312]: Why is this still needed? if a user has specified a format then that should be used always. So just returning `this.replace(target, scopedVars, 'text');` should be enough

**Review by torkelo — COMMENTED** [review]
Looks good!, still some unnecessary parsing I think  replaceWithText

File: public/app/features/templating/template_srv.ts (hunk 11 above)
Comment (mckn) [reply public/app/features/templating/template_srv.ts:312]: Or should we change that logic?

Comment (torkelo) [reply public/app/features/templating/template_srv.ts:312]: yea, think that one is not correct (or can be changed). 

The use case for replaceWithText is really only in panel title, and legend / display name inputs. And the reason for all the other formats are to work with query syntaxes. So if someone want's glob formatting then that's what they get.

File: public/app/features/templating/template_srv.ts [hunk 12]
```diff
@@ -248,8 +265,11 @@ export class TemplateSrv implements BaseTemplateSrv {
       }
 
       let value = variable.current.value;
+      let text = variable.current.text;
+
       if (this.isAllValue(value)) {
         value = this.getAllValue(variable);
+        text = 'All';
```
Comment (hugohaggmark) [inline public/app/features/templating/template_srv.ts:272]: nitpick: I think there is a `const` for this

**Review by hugohaggmark — COMMENTED** [review]
Looks the best, great that you changed the gdev-dash, docs and e2e!! 🌟

**Review by torkelo — APPROVED** [review]
Looks great!

**Pull request merged**